wagtail/contrib/wagtailapi/serializers.py

```diff
@@ -2,171 +2,268 @@
-                    ])
-                else:
-                    yield field_name, None
-            else:
-                yield field_name, field._get_val_from_obj(obj)
-
-            continue
-        except models.fields.FieldDoesNotExist:
-            pass
-
-        # Check attributes
-        if hasattr(obj, field_name):
-            value = getattr(obj, field_name)
-            yield field_name, force_text(value, strings_only=True)
-            continue
-
-
-class WagtailSerializer(BaseSerializer):
-    def to_representation(self, instance):
-        request = self.context['request']
-        fields = self.context.get('fields', frozenset())
-        all_fields = self.context.get('all_fields', False)
-        show_details = self.context.get('show_details', False)
-        return self.serialize_object(
-            request,
-            instance,
-            fields=fields,
-            all_fields=all_fields,
-            show_details=show_details
-        )
-
-    def serialize_object_metadata(self, request, obj, show_details=False):
-        """
-        This returns a JSON-serialisable dict to use for the "meta"
-        section of a particlular object.
-        """
-        data = OrderedDict()
-
-        # Add type
-        data['type'] = type(obj)._meta.app_label + '.' + type(obj).__name__
-        data['detail_url'] = ObjectDetailURL(type(obj), obj.pk)
+    "meta": {
+        "type": "wagtailimages.Image",
+        "detail_url": "http://api.example.com/v1/images/1/"
+    }
+    """
+    def get_attribute(self, instance):
+        return instance
+
+    def to_representation(self, obj):
+        return OrderedDict([
+            ('type', type(obj)._meta.app_label + '.' + type(obj).__name__),
+            ('detail_url', ObjectDetailURL(type(obj), obj.pk)),
+        ])
+
+
+class PageMetaField(MetaField):
+    """
+    A subclass of MetaField for Page objects.
+
+    Changes the "type" field to use the name of the specific model of the page.
+    """
+    def to_representation(self, page):
+        data = super(PageMetaField, self).to_representation(page)
+
+        # Change type to the specific page class instead
+        data['type'] = page.specific_class._meta.app_label + '.' + page.specific_class.__name__
 
         return data
 
-    def serialize_object(self, request, obj, fields=frozenset(), extra_data=(), all_fields=False, show_details=False):
-        """
-        This converts an object into JSON-serialisable dict so it can
-        be used in the API.
-        """
-        data = [
-            ('id', obj.id),
+
+class DocumentMetaField(MetaField):
+    """
+    A subclass of MetaField for Document objects.
+
+    Adds a "download_url" field.
+    """
+    def to_representation(self, document):
+        data = super(DocumentMetaField, self).to_representation(document)
+
+        # Add download url
+        if self.context.get('show_details', False):
+            data['download_url'] = URLPath(document.url)
```

Comment:
Same again - I'd consider just overriding the whole method rather than calling super.
Also to consider - is `URLPath` useful or should it just return the hyperlink as a string at this point?